src/main/scala/is/hail/utils/ArrayStack.scala

```diff
@@ -0,0 +1,50 @@
+package is.hail.utils
+
+import scala.reflect.ClassTag
+
+final class ArrayStack[@specialized T](hintSize: Int = 16)(implicit tct: ClassTag[T]) {
+  private var a = new Array[T](hintSize)
+  
+  private var size_ = 0
+  
+  def size: Int = size_
+
+  def capacity: Int = a.length
+  
+  def isEmpty: Boolean = size_ == 0
+
+  def nonEmpty: Boolean = size_ > 0
+
+  def top: T = {
+    assert(size_ > 0)
+    a(size_ - 1)
+  }
+
+  def push(x: T) {
+    if (size_ == a.length) {
+      val newA = new Array[T](size_ * 2)
+      System.arraycopy(a, 0, newA, 0, size_)
+      a = newA
+    }
+    a(size_) = x
+    size_ += 1
+  }
+
+  def pop(): T = {
+    assert(size_ > 0)
+    size_ -= 1
+    val x = a(size_)
+    a(size_) = uninitialized[T]
```

Comment:
I think this is unnecessary. The privacy of a and the assert on `apply` makes the value at this position effective unknowable.

Reply:
This is needed to let popped stack elements get garbage collected.

Reply:
won't this box?

Reply:
I made uninitialized (which was still allocating?!) specialized.